Refactor printers

clinssen opened the pull request:
Clean up printers along the "compositionality" pattern to facilitate adding other target platforms in a flexible way.

"[...] a delegator visitor has a setter method for visitors of each (transitive) parent grammar" [2] This suggests a "clean" way of dividing up the responsibility of the printers: along grammar rule boundaries.

[1] A. Butting, K. Hölldobler B. Rumpe, A. Wortmann: Compositional Modelling Languages with Analytics and Construction Infrastructures Based on Object-Oriented Techniques—The MontiCore Approach. In: Composing Model-Based Analysis Tools, pp. 217-234, Springer, 2021
[2] Robert Heim, Pedram Mir Seyed Nazari1, Bernhard Rumpe and Andreas Wortmann. Compositional Language Engineering Using Generated, Extensible, Static Type-Safe Visitors. Lecture Notes in Computer Science, Andrzej Wąsowski, Henrik Lönn (eds.) - Modelling Foundations and Applications, 12th European Conference, ECMFA 2016

## Class diagram

Compositionality pattern; note the loop between ExpressionPrinter -> SimpleExpressionPrinter -> FunctionCallPrinter.

[classes.pdf](https://github.com/nest/nestml/files/10184603/classes.pdf)

# TODO

- [x] If a numeric solver is used, currently all state variables are moved into a state vector. However, non-ODE variables (such as a refractory counter of type int) should not be part of those ODEs; they are only written to in the ``update`` block. *Done as part of #844.*
- [x] replace ``print_nestml_type()`` methods on ``TypeSymbol``s with a new ``NESTMLTypesPrinter`` *Moved to a new issue: #845*

pnbabu commented on pynestml/codegeneration/resources_nest/point_neuron/common/SynapseHeader.h.jinja2:
```diff
@@ -567,18 +552,18 @@ public:
 {#
```
Is the commented code needed?

pnbabu commented on pynestml/codegeneration/resources_nest/point_neuron/directives/MemberInitialization.jinja2:
```diff
@@ -1,19 +1,20 @@
-  {{printer.print_origin(variable)}}{{names.name(variable)}} = 0; // as {{variable.get_type_symbol().print_symbol()}}
+  {{printer.print(variable)}} = 0; // as {{sym.get_type_symbol().print_symbol()}}
   {%- endif -%}
 {%- endif -%}
```
Please add a new line

pnbabu commented on pynestml/codegeneration/resources_nest/point_neuron/directives/MemberVariableGetterSetter.jinja2:
```diff
@@ -1,17 +1,16 @@
 }
-
-{% endif -%}
+{%- endif %}
```
Please add a new line

pnbabu commented on pynestml/meta_model/ast_neuron_or_synapse.py:
```diff
@@ -433,6 +435,32 @@ def get_parameter_invariants(self):
+
+    def get_state_variables(self) -> List[ASTVariable]:
+        """
+        Returns a list of all parameters.
```
```suggestion
        Returns a list of all state vairables.
```

pnbabu commented on pynestml/utils/ast_utils.py:
```diff
@@ -1316,17 +1318,114 @@ def get_input_port_by_name(cls, input_blocks: List[ASTInputBlock], port_name: st
+    @classmethod
+    def get_parameter_variable_by_name(cls, node: ASTNeuronOrSynapse, var_name: str) -> ASTVariable:
+        """
+        Get the declaration based on the name of the parameter
+        :param parameters_block: the parameter block
+        :param var_name: variable name to be searched
+        :return: declaration containing the variable
```
```suggestion
        Get the parameter variable based on the name of the parameter
        :param node: the neuron or synapse
        :param var_name: variable name to be searched
        :return: the parameter variable
```

pnbabu commented on pynestml/utils/ast_utils.py:
```diff
@@ -1316,17 +1318,114 @@ def get_input_port_by_name(cls, input_blocks: List[ASTInputBlock], port_name: st
+    @classmethod
+    def get_internal_by_name(cls, node: ASTNeuronOrSynapse, var_name: str) -> ASTDeclaration:
+        """
+        Get the declaration based on the name of the parameter
```
Could you change the docstrings of the `get_` functions in this file to match the function definition?

pnbabu reviewed (commented):
This is huge with a lot of refactoring! Thanks for all the work. I have tried to review as best as I can. Please find my comments.

clinssen commented on pynestml/codegeneration/nest_code_generator.py:
```diff
@@ -124,36 +136,62 @@ def __init__(self, options: Optional[Mapping[str, Any]] = None):
+                                                                                                             constant_printer=self._constant_printer,
+                                                                                                             function_call_printer=self._nest_function_call_printer_no_origin))
+        self._nest_variable_printer_no_origin._expression_printer = self._printer_no_origin
+        self._nest_function_call_printer_no_origin._expression_printer = self._printer_no_origin
```
Comments added

clinssen commented on pynestml/codegeneration/printers/cpp_expression_printer.py:
```diff
@@ -19,140 +19,221 @@
+    def print_expression(self, node: ASTExpressionNode) -> str:
+        assert isinstance(node, ASTExpression)
+        if node.is_unary_operator():
+            return self.print_unary_op(node)
```
Renamed as suggested and made "private" methods (prefixed with underscore)

clinssen commented on pynestml/codegeneration/printers/cpp_expression_printer.py:
```diff
@@ -19,140 +19,221 @@
+        rhs = self.print(node.get_expression())
+        unary_operator = node.get_unary_operator()
+        if unary_operator.is_unary_plus:
+            return "(" + "+" + rhs + ")"
```
Extra safety to preserve the intended operator order of application.

clinssen commented on pynestml/codegeneration/printers/cpp_printer.py:
```diff
@@ -64,116 +60,143 @@
 
 
-class NestPrinter(Printer):
+class CppPrinter(ASTPrinter):
```
``CppPrinter`` contains only generic C++ syntax components irrespective of NEST.

clinssen commented on pynestml/codegeneration/printers/cpp_variable_printer.py:
```diff
@@ -0,0 +1,54 @@
+        :param node: a single variable symbol or variable
+        :return: a string representation
+        """
+        assert isinstance(node, ASTVariable)
```
Ideally yes, hopefully Python will get a runtime type check feature at some stage. For now we can maybe keep this as this kind of check appears in quite a number of functions (but admittedly inconsistently).

clinssen commented on pynestml/codegeneration/printers/types_printer.py:
```diff
@@ -21,15 +21,16 @@
 
 
-class TypesPrinter(metaclass=abc.ABCMeta):
+class TypesPrinter(SymbolPrinter, metaclass=abc.ABCMeta):
```
Renamed ``TypesPrinter`` to ``TypeSymbolPrinter`` to better indicate the class hierarchy.

clinssen commented on pynestml/codegeneration/resources_nest/point_neuron/common/NeuronClass.jinja2:
```diff
@@ -395,18 +397,27 @@ void {{neuronName}}::recompute_internal_variables(bool exclude_timestep) {
-{%- if variable.name != "__h" %}
-{%-   include "directives/MemberInitialization.jinja2" %}
-{%- endif %}
+{%- for internals_block in neuron.get_internals_blocks() %}
```
Moving from passing symbols to passing ``ASTNode``s due to the latter being preferred (see earlier comment)

clinssen commented:
Thank you for the detailed review! Where the updates immediately addressed your comments, I have already marked them as resolved. The others I have commented upon. Could I ask you to take another look?

clinssen commented:
Merge only after #846.

pnbabu commented on pynestml/codegeneration/printers/latex_expression_printer.py:
```diff
@@ -104,29 +74,79 @@ def __do_print(self, node: ASTExpressionNode, prefix="") -> str:
 
-    def print_function_call(self, function_call: ASTFunctionCall) -> str:
-        function_name = self.reference_converter.convert_function_call(function_call)
+    def _print_unary_operator(self, ast_unary_operator) -> str:
```
The function names need to be changed here also with the suffix `_expression`

pnbabu commented on pynestml/codegeneration/printers/nest_variable_printer.py:
```diff
@@ -0,0 +1,150 @@
+            return ""
+
+        vector_param = ""
+        if self.with_vector_parameter and symbol.has_vector_parameter():
```
Why is `with_vector_parameter` necessary?

clinssen replied:
It's because in some cases, a vector variable needs to be printed without the expression in square brackets, where in other cases the size or index parameter in brackets in necessary. An alternative might have been to have a variant of the ``NESTVariablePrinter`` that prints without, and one that prints with this part.

pnbabu commented on pynestml/codegeneration/printers/ode_toolbox_expression_printer.py:
```diff
@@ -0,0 +1,71 @@
+
+        return "Piecewise((" + if_true + ", " + condition + "), (" + if_not + ", True))"
+
+    def _print_arithmetic_operator(self, node: ASTExpressionNode) -> str:
```
```suggestion
    def _print_arithmetic_operator_expression(self, node: ASTExpressionNode) -> str:
```

pnbabu commented on pynestml/codegeneration/printers/simple_expression_printer.py:
```diff
@@ -0,0 +1,67 @@
+
+    @abstractmethod
+    def print_simple_expression(self, node: ASTSimpleExpression) -> str:
+        """Print an expression.
+
+        Parameters
+        ----------
+        node : ASTExpressionNode
+            The expression node to print.
+
```
```suggestion
        """Print a simple expression.

        Parameters
        ----------
        node : ASTSimpleExpression
            The simple expression to print.

```

pnbabu commented on pynestml/codegeneration/printers/type_symbol_printer.py:
```diff
@@ -0,0 +1,36 @@
+from pynestml.codegeneration.printers.symbol_printer import SymbolPrinter
+
+
+class TypeSymbolPrinter(SymbolPrinter, metaclass=abc.ABCMeta):
```
Why does `TypeSymbolPrinter` need to inherit from `SymbolPrinter`? Since both have the same `print` function and the `SymbolPrinter` is not used anywhere else, can we eliminate it altogether?

clinssen replied:
You are correct, but I'd like to keep this organization, because in principle we might also be printing other symbols (like ``FunctionSymbol``s).

pnbabu reviewed (commented):
I have some minor comments

pnbabu reviewed (approved):
Great, thanks!
